cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/controller/EndpointsController.kt

```diff
@@ -0,0 +1,114 @@
+package com.cognifide.cogboard.config.controller
+
+import com.cognifide.cogboard.CogboardConstants
+import com.cognifide.cogboard.config.EndpointLoader
+import com.cognifide.cogboard.config.EndpointLoader.Companion.ENDPOINTS
+import com.cognifide.cogboard.config.EndpointLoader.Companion.ENDPOINT_ID_PREFIX
+import com.cognifide.cogboard.config.EndpointLoader.Companion.ENDPOINT_ID_PROP
+import com.cognifide.cogboard.config.EndpointLoader.Companion.ENDPOINT_LABEL_PROP
+import com.cognifide.cogboard.storage.Storage
+import com.cognifide.cogboard.storage.docker.VolumeStorage
+import io.vertx.core.AbstractVerticle
+import io.vertx.core.json.JsonArray
+import io.vertx.core.json.JsonObject
+
+class EndpointsController : AbstractVerticle() {
+
+    private lateinit var storage: Storage
+
+    override fun start() {
+        storage = VolumeStorage(vertx)
+        listenOnEndpointsUpdate()
+        listenOnEndpointsDelete()
+    }
+
+    private fun listenOnEndpointsUpdate() = vertx
+            .eventBus()
+            .consumer<JsonObject>(CogboardConstants.EVENT_UPDATE_ENDPOINTS_CONFIG)
+            .handler {
+                save(it.body())
+                storage.saveEndpointsConfig(config())
+            }
+
+    private fun listenOnEndpointsDelete() = vertx
+            .eventBus()
+            .consumer<JsonObject>(CogboardConstants.EVENT_DELETE_ENDPOINTS_CONFIG)
+            .handler {
+                delete(it.body())
+                storage.saveEndpointsConfig(config())
+            }
+
+    fun save(endpoint: JsonObject) = if (exists(endpoint)) update(endpoint) else add(endpoint)
+
+    fun exists(endpoint: JsonObject): Boolean {
+        val endpointId = endpoint.getString(ENDPOINT_ID_PROP) ?: return false
+        val endpoints = config().getJsonArray(ENDPOINTS) ?: JsonArray()
+        return endpoints.stream()
+                .map { it as JsonObject }
+                .anyMatch {
+                    endpointId == it.getString(ENDPOINT_ID_PROP)
+                }
+    }
+
+    private fun update(endpoint: JsonObject) {
+        val endpointId = endpoint.getString(ENDPOINT_ID_PROP)
+        val endpointToUpdate = EndpointLoader.from(config(), endpointId).load()
+        endpointToUpdate.mergeIn(endpoint, true)
+    }
+
+    private fun add(endpoint: JsonObject) {
+        endpoint.let {
+            it.put(ENDPOINT_ID_PROP, generateId())
+            it.put(ENDPOINT_LABEL_PROP, it.getString(ENDPOINT_LABEL_PROP) ?: it.getString(ENDPOINT_ID_PROP))
+        }
+
+        config()
+                .putIfNotExist(ENDPOINTS, JsonArray())
```

Comment:
I got rid of this nasty **if** **statement** and defined custom extension function as seen below.